jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,170 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingNestedCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link Content.Sink} backed by another {@link Content.Sink}.
+ * Any content written to this {@link Content.Sink} is buffered,
+ * then written to the delegate's
+ * {@link Content.Sink#write(boolean, ByteBuffer, Callback)}. </p>
+ */
+public class BufferedContentSink implements Content.Sink
+{
+    private final Content.Sink _delegate;
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxBufferSize;
+    private CountingByteBufferAccumulator _accumulator;
+    private boolean _firstWrite = true;
+    private boolean _lastWritten;
+
+    public BufferedContentSink(Content.Sink delegate, ByteBufferPool bufferPool, boolean direct, int maxBufferSize)
+    {
+        if (maxBufferSize <= 0)
+            throw new IllegalArgumentException("maxBufferSize must be > 0, was: " + maxBufferSize);
+        _delegate = delegate;
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+    {
+        if (_lastWritten)
+        {
+            callback.failed(new IOException("complete"));
+            return;
+        }
+        if (_firstWrite)
+        {
+            _accumulator = new CountingByteBufferAccumulator(_bufferPool, _direct, _maxBufferSize);
+            _firstWrite = false;
+        }
+        _lastWritten |= last;
+
+        ByteBuffer current = byteBuffer != null ? byteBuffer : BufferUtil.EMPTY_BUFFER;
+        IteratingNestedCallback writer = new IteratingNestedCallback(callback)
```

Comment:
I don't think the use of an iterating callback is optimal here. Consider the case of an app writing 1 byte at a time, every write will create an new iterating callback and yet still call the callback within the scope of `iterate()`, which might do another write and make a new iterating callback etc. etc.    So we still get stack overflow and create a lot of garbage

What I think we need is something like:
 + a write comes in that is smaller that the left over space in the aggregate buffer
 + copy the data to the buffer and succeed the callback
     - another write comes in, but we see that we are still calling the previous callback so:
     - we copy the data to the aggregate buffer and return without completing the callback
 + we return from the succeeding of the original callback and see that we have another write in progress so we complete its callback

Of course that is more complex when we have writes that are bigger than the buffer.... BUT I think for any write larger that a particular size, we should not aggregate it and just write it directly.  See the logic regarding `_commitSize` in `HttpOutput`